src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/WindowsForms/WindowsFormsEditorProvider.cs

```diff
@@ -0,0 +1,176 @@
+    /// </summary>
+    [Export(typeof(IProjectSpecificEditorProvider))]
+    [AppliesTo(ProjectCapability.DotNet)]
+    [Order(Order.BeforeDefault)] // Need to run before CPS's version before its deleted
+    internal partial class WindowsFormsEditorProvider : IProjectSpecificEditorProvider
+    {
+        private readonly static SubTypeDescriptor[] SubTypeDescriptors = new[]
+        {
+            new SubTypeDescriptor("Form",           VSResources.WindowsFormEditor_DisplayName, "Form",         useDesignerByDefault: true), // "Form" and "Designer" represent the same thing
+            new SubTypeDescriptor("Designer",       VSResources.WindowsFormEditor_DisplayName, "Form",         useDesignerByDefault: true),
+            new SubTypeDescriptor("UserControl",    VSResources.UserControlEditor_DisplayName, "UserControl",  useDesignerByDefault: true),
+            new SubTypeDescriptor("Component",      VSResources.ComponentEditor_DisplayName,   "Component",    useDesignerByDefault: false)
+        };
+
+        private readonly ActiveConfiguredProject<ProjectProperties> _properties;
+        private readonly Lazy<IPhysicalProjectTree> _projectTree;
+        private readonly Lazy<IProjectSystemOptions> _options;
+        private readonly IVsService<IVsFeatureFlags> _featureFlagsService;
+
+        [ImportingConstructor]
+        public WindowsFormsEditorProvider(UnconfiguredProject unconfiguredProject, ActiveConfiguredProject<ProjectProperties> properties, Lazy<IPhysicalProjectTree> projectTree, Lazy<IProjectSystemOptions> options, IVsService<SVsFeatureFlags, IVsFeatureFlags> featureFlagsService)
+        {
+            _properties = properties;
+            _projectTree = projectTree;
+            _options = options;
+            _featureFlagsService = featureFlagsService;
+
+            ProjectSpecificEditorProviders = new OrderPrecedenceImportCollection<IProjectSpecificEditorProvider, INamedExportMetadataView>(projectCapabilityCheckProvider: unconfiguredProject);
+        }
+
+        [ImportMany]
+        public OrderPrecedenceImportCollection<IProjectSpecificEditorProvider, INamedExportMetadataView> ProjectSpecificEditorProviders
+        {
+            get;
+        }
+
+        public async Task<IProjectSpecificEditorInfo?> GetSpecificEditorAsync(string documentMoniker)
+        {
+            Requires.NotNullOrEmpty(documentMoniker, nameof(documentMoniker));
+
+            IProjectSpecificEditorInfo? editor = await GetDefaultEditorAsync(documentMoniker);
+            if (editor == null)
+                return null;
+
+            SubTypeDescriptor? descriptor = await GetSubTypeDescriptorAsync(documentMoniker);
+            if (descriptor == null)
+                return null;
+
+            // Project properties are more expensive than above, so only
+            // read them if absolutely must open the designer for the file
+            if (!await IsDesignerEnabledAsync())
+                return null;
```

Comment:
This method serialises three async operations. Would it reduce delay when opening a file if they were all kicked off at the same time before being awaited? Based on your comment you may wish to do this only for the first two.